### ISSUE TITLE
Implement General Activity Dialog

### ISSUE THREAD

**Issue by wbaccinelli** [PerfectFit-project/virtual-coach-server#290]
Implement the C1.16 General Activity Dialog as described [here](https://teams.microsoft.com/_#/xlsx/viewer/aggregatefiles/https:~2F~2Fleidenuniv1.sharepoint.com~2Fsites~2FTeamAio~2FShared%2520Documents~2FGeneral~2FContent%2520folder~2FVirtual%2520Coach%2520Dialogs%2520and%2520Resources~2FC1%2520Dialogs~2F260922%2520C1.16%2520General%2520Activity%2520Dialogue.xlsx?baseUrl=https:~2F~2Fleidenuniv1.sharepoint.com~2Fsites~2FTeamAio&fileId=141a450d-7c56-408e-abb2-312cadd43f92&ctx=aggregate&viewerAction=view)

**Comment by Wouter1** [comment]
Checked with Walter.
We put the edits on a new branch 290-general-activity
We create new files stories_general_activity.yml and domain_general_activity etc as needed

[260922 C1.16 General Activity Dialogue.xlsx](https://github.com/PerfectFit-project/virtual-coach-issues/files/9724117/260922.C1.16.General.Activity.Dialogue.xlsx)

Here is the dialog to be implemented

build instructions:
```
source venv/bin/activate 
cd Rasa_Bot
rasa train --domain domain/
docker compose up --build

curl --location --request POST 'http://localhost:5005/webhooks/rest/webhook' --header 'Content-Type text/plain' --data-raw '{ "sender":"41538", "message":"hallo" } '
```

**Comment by Wouter1** [comment]
The story needs access to the FirstAidKit.
FirstAidKit is a database object defined in virtual-coach-db/dbschema/models.py

**Comment by Wouter1** [comment]
Checked the validate_see_myself_as_smoker_form action to see how numbers are asked.

The action returns a dict {'see_myself_as_picked_words_smoker":N}. N may be a number of "None"

I dont see anyone using this action so it might not be working code 

In other code I see a call to SlotSet. But if we would do that, the action would not be reusable for other actions also needing a number

**Comment by Wouter1** [comment]
@wbaccinelli I'm a bit puzzled by the return value of the actions.
If I get it correct, Actions must return a dict with slot name and value.

I'm not sure but it looks like you can return
```{"name":slotname. "value":slotvalue, "timestamp":floatval, "event":"slot"}```
or just
```{"confirm_words_response": yes_or_no_response}```

What happens in the latter case, is this a shortcut for an event of type slot?

Is a return from an Action always resulting in something put into a slot?

Can we make a general action that asks a number between 0 and 10?
And can we then later move the resulting number into the proper slot?
Or, can we pass the slot name as argument into the action?

**Comment by BendikC** [comment]
@Wouter1 I hope I can answer your questions to some extent. For the question you had about the forms rasa has a built in function for forms, where they create a loop and keep asking for the input until the slot is filled. The code that you see is returning an empty slot because the input is invalid, as seen by the method _is_valid_input and so the dispatcher line actually tells rasa to give this response back to the user "utter_please_answer_1_2_3", then the return: none tells rasa that the slot is empty. Since the form requires the slot to be filled, this triggers the loop so that rasa asks for the input again. This will go on for as long as the user gives an invalid input, and once the input is valid it exits the form.

The question about the slots is a bit confusing to me. return SlotSet["slotname", "slot_value"]  specifies for the slot "slotname" that we associate the specific "slotvalue" to it. We can define as many slots as we need to store different types of information, so if we store a number into one slot we can store a different number into a different slot, or we could overwrite the number in the slot to a different one using the same command. We can also have text, booleans, categorical slots etc...

For your final one about actions, actions do not need to return anything. An action is simply a function that is triggered by rasa. An example of this is in actions_minimal_functional_product on line 177 where an empty array is returned. {"confirm_words_response": yes_or_no_response} is a shortcut for setting the slot to the specific value, I dont think there are any major differences between this and SlotSet, as they should provide the same functionality. 
For a general action asking for a number between 1-10, I am not completely sure how we could generalize this but one possibility would be to make a form that asks for a number between 1-10, store a slot that gives you the slot name of the slot to be filled then in the dialog itself change the slotname of this slot to the one corresponding to our dialog and then call the general action, if this makes sense.

**Comment by Wouter1** [comment]
@BendikC  thanks!

**Comment by Wouter1** [comment]
General usage and test procedure at this moment:
```
rasa train --domain domain/
docker compose up
```
and then in another window execute messages
```
curl --location --request POST 'http://localhost:5005/webhooks/rest/webhook' --header 'Content-Type text/plain' --data-raw '{ "sender":"41538", "message":"hallo" } '
```

**Comment by Wouter1** [comment]
>Hoe ging de laatste activiteit?

gives now

>Wil jij deze oefening nu doen of later?

Strange, where does that come from and did I make a mistake?

**Comment by Wouter1** [comment]
That reply comes from domain_rescheduling_dialog
I suppose this comes from checkpoint: rescheduling_query_dialog  inside my "introduction PF" story?
But I first do an utter_greed? Why does that not appear first?

**Comment by Wouter1** [comment]
Added to domain_general_activity
```
intents:
  - review_last_activity
```

**Comment by Wouter1** [comment]
That did nothing

**Comment by Wouter1** [comment]
Did global search for entities in the code, gives nothing. Are we not using rasa entities?

**Comment by Wouter1** [comment]
Mailed Walter to ask if he has some time to discuss

**Comment by Wouter1** [comment]
If I add "Bla" to review_last_activity examples in nlu, I DO get th the correct phase and can do some testing

**Comment by Wouter1** [comment]
[https://rasa.com/docs/rasa/stories/#checkpoints checkpoint documentation on rasa] is really poor as they only give an example and do not explain anything how it works
Only clue I see is on this ticket https://forum.rasa.com/t/solved-how-to-use-check-points-in-story/1351/3 where akelad says "If one story ends with a checkpoint, and another story starts with a checkpoint of the same name, that’s where the conversation will go". So the *location* where you put the checkpoint makes a huge difference

**Comment by Wouter1** [comment]
There seems problem with my ValidateActivityUsefulnessForm. I can not seem to call it properly.
```
2022-10-03 12:49:14 ERROR    rasa.engine.graph  - Error running graph component for node train_RulePolicy1.
ActionNotFoundException: Cannot access action 'validate_activity_usefulness_form', as that name is not a registered action for this domain. Available actions are: 
	 - action_listen
	 - action_restart
	 - action_session_start
    .........
```

I am extending FormValidationAction 
But I do not understand how the "validate_rescheduling_option" function works. Who is calling this?

**Comment by Wouter1** [comment]
Strange, I can call the action validate_activity_usefulness_form from the stories no problem. But if I do active_loop on that same validate_activity_usefulness_form I get above error

**Comment by Wouter1** [comment]
I now get ```rasa_bot-rasa_server-1   | /usr/local/lib/python3.8/site-packages/rasa/shared/utils/io.py:98: UserWarning: Parsed an intent 'review_last_activity' which is not defined in the domain. Please make sure all intents are listed in the domain.```. But in the domain_general_activity there IS 
```
intents:
  - review_last_activity
```


Before that error I also see
````
rasa_bot-rasa_actions-1  | Exception occurred while handling uri: 'http://rasa_actions:5055/webhook'
rasa_bot-rasa_actions-1  | Traceback (most recent call last):
rasa_bot-rasa_actions-1  |   File "handle_request", line 83, in handle_request
rasa_bot-rasa_actions-1  |     )
rasa_bot-rasa_actions-1  |   File "/opt/venv/lib/python3.8/site-packages/rasa_sdk/endpoint.py", line 104, in webhook
rasa_bot-rasa_actions-1  |     result = await executor.run(action_call)
rasa_bot-rasa_actions-1  |   File "/opt/venv/lib/python3.8/site-packages/rasa_sdk/executor.py", line 397, in run
rasa_bot-rasa_actions-1  |     events = await utils.call_potential_coroutine(
rasa_bot-rasa_actions-1  |   File "/opt/venv/lib/python3.8/site-packages/rasa_sdk/utils.py", line 230, in call_potential_coroutine
rasa_bot-rasa_actions-1  |     return await coroutine_or_return_value  # type: ignore [misc] # https://github.com/python/mypy/issues/7587
rasa_bot-rasa_actions-1  |   File "/app/actions/actions_minimum_functional_product.py", line 70, in run
rasa_bot-rasa_actions-1  |     session = get_db_session(db_url=DATABASE_URL)
rasa_bot-rasa_actions-1  |   File "/opt/venv/lib/python3.8/site-packages/virtual_coach_db/helper/helper_functions.py", line 25, in get_db_session
rasa_bot-rasa_actions-1  |     sanitized_db_url = santize_db_url(db_url)
rasa_bot-rasa_actions-1  |   File "/opt/venv/lib/python3.8/site-packages/virtual_coach_db/helper/helper_functions.py", line 18, in santize_db_url
rasa_bot-rasa_actions-1  |     if db_url.startswith("postgres://"):
rasa_bot-rasa_actions-1  | AttributeError: 'NoneType' object has no attribute 'startswith'

```

**Comment by Wouter1** [comment]
Ah there is yaml syntax  error with the domain_general_activity file
```
/home/wouter/git/virtual-coach-main/venv/lib/python3.8/site-packages/rasa/shared/core/domain.py:1804: UserWarning: The file /home/wouter/git/virtual-coach-rasa/Rasa_Bot/domain/domain_general_activity.yml could not be loaded as domain file. You can use https://yamlchecker.com/ to validate the YAML syntax of your file.
```
It is strange that this is shown with dark orange and as warning only. This seems fatal and rasa should stop on it

**Comment by Wouter1** [comment]
I had a slot partially commented out.
After fixing things however seem to break further. When I send message to rasa, nothing happens except the usual "unable to open database file". Maybe this error is the problem
>rasa_bot-rasa_server-1   | /usr/local/lib/python3.8/site-packages/rasa/shared/utils/io.py:98: UserWarning: No core model. Skipping action prediction and execution.

But I did not see errors in the build?

**Comment by Wouter1** [comment]
Ok, this is weird
If you do NOT set "version:3.1" at the top of the yml files, you get warnings in the build process.
If you DO set the version, intents in the ym files are not recognised 

It might be related to indentation levels in the yml file. There seems a difference between 

```
intents:
- review_last_activity

```

and

```
intents:
  - review_last_activity
```

**Comment by Wouter1** [comment]
Stange, even if you specify the version in just 1 file it seems to take that as default for the other files.

**Comment by Wouter1** [comment]
Getting
>/home/wouter/git/virtual-coach-main/venv/lib/python3.8/site-packages/rasa/shared/utils/io.py:98: UserWarning: Loading domain from 'domain/' failed. Using empty domain. Error: 'You need to specify a value for the key 'value' in the slot mapping of type 'from_intent' for slot 'activity_useful_rating'. Please see https://rasa.com/docs/rasa/domain#slots for more information.'


This warning is in fact an error and it's not clear that it comes from parsing domain_general_activity.yml...

**Comment by Wouter1** [comment]
I added this to  the first steps of the review_last_activity story:

>    - action: action_get_name_from_database

This action fails with 
```
rasa_bot-rasa_server-1   | 2022-10-03 12:23:30 ERROR    rasa.core.tracker_store  - Error happened when trying to save conversation tracker to 'SQLTrackerStore'. Falling back to use the 'InMemoryTrackerStore'. Please investigate the following error: (sqlite3.OperationalError) unable to open database file
rasa_bot-rasa_server-1   | (Background on this error at: https://sqlalche.me/e/14/e3q8).
rasa_bot-rasa_actions-1  | Exception occurred while handling uri: 'http://rasa_actions:5055/webhook'
rasa_bot-rasa_actions-1  | Traceback (most recent call last):
rasa_bot-rasa_actions-1  |   File "handle_request", line 83, in handle_request
rasa_bot-rasa_actions-1  |     )
rasa_bot-rasa_actions-1  |   File "/opt/venv/lib/python3.8/site-packages/rasa_sdk/endpoint.py", line 104, in webhook
rasa_bot-rasa_actions-1  |     result = await executor.run(action_call)
rasa_bot-rasa_actions-1  |   File "/opt/venv/lib/python3.8/site-packages/rasa_sdk/executor.py", line 397, in run
rasa_bot-rasa_actions-1  |     events = await utils.call_potential_coroutine(
rasa_bot-rasa_actions-1  |   File "/opt/venv/lib/python3.8/site-packages/rasa_sdk/utils.py", line 230, in call_potential_coroutine
rasa_bot-rasa_actions-1  |     return await coroutine_or_return_value  # type: ignore [misc] # https://github.com/python/mypy/issues/7587
rasa_bot-rasa_actions-1  |   File "/app/actions/actions_minimum_functional_product.py", line 70, in run
rasa_bot-rasa_actions-1  |     session = get_db_session(db_url=DATABASE_URL)
rasa_bot-rasa_actions-1  |   File "/opt/venv/lib/python3.8/site-packages/virtual_coach_db/helper/helper_functions.py", line 25, in get_db_session
rasa_bot-rasa_actions-1  |     sanitized_db_url = santize_db_url(db_url)
rasa_bot-rasa_actions-1  |   File "/opt/venv/lib/python3.8/site-packages/virtual_coach_db/helper/helper_functions.py", line 18, in santize_db_url
rasa_bot-rasa_actions-1  |     if db_url.startswith("postgres://"):
rasa_bot-rasa_actions-1  | AttributeError: 'NoneType' object has no attribute 'startswith'
rasa_bot-rasa_server-1   | 2022-10-03 12:23:30 ERROR    rasa.core.processor  - Encountered an exception while running action 'action_get_name_from_database'.Bot will continue, but the actions events are lost. Please check the logs of your action server for more information.
rasa_bot-rasa_server-1   | Traceback (most recent call last):
rasa_bot-rasa_server-1   |   File "/usr/local/lib/python3.8/site-packages/rasa/core/actions/action.py", line 741, in run
rasa_bot-rasa_server-1   |     response: Any = await self.action_endpoint.request(
rasa_bot-rasa_server-1   |   File "/usr/local/lib/python3.8/site-packages/rasa/utils/endpoints.py", line 173, in request
rasa_bot-rasa_server-1   |     raise ClientResponseError(
rasa_bot-rasa_server-1   | rasa.utils.endpoints.ClientResponseError: 500, Internal Server Error, body='b'{"description":"Internal Server Error","status":500,"message":"The server encountered an internal error and cannot complete your request."}''
rasa_bot-rasa_server-1   | 
rasa_bot-rasa_server-1   | The above exception was the direct cause of the following exception:
rasa_bot-rasa_server-1   | 
rasa_bot-rasa_server-1   | Traceback (most recent call last):
rasa_bot-rasa_server-1   |   File "/usr/local/lib/python3.8/site-packages/rasa/core/processor.py", line 874, in _run_action
rasa_bot-rasa_server-1   |     events = await action.run(
rasa_bot-rasa_server-1   |   File "/usr/local/lib/python3.8/site-packages/rasa/core/actions/action.py", line 765, in run
rasa_bot-rasa_server-1   |     raise RasaException(
rasa_bot-rasa_server-1   | rasa.shared.exceptions.RasaException: Failed to execute custom action 'action_get_name_from_database'
rasa_bot-rasa_server-1   | 2022-10-03 12:23:30 ERROR    rasa.core.tracker_store  - Error happened when trying to save conversation tracker to 'SQLTrackerStore'. Falling back to use the 'InMemoryTrackerStore'. Please investigate the following error: (sqlite3.OperationalError) unable to open database file
rasa_bot-rasa_server-1   | (Background on this error at: https://sqlalche.me/e/14/e3q8).

```

I think because there is no known user name in the database.

**Comment by Wouter1** [comment]
Somehow we are not jumping to the rate_usefulness checkpoint
Maybe because my first story did not have a checkpoint for itself? That may have caused the checkpoint "call" that I thought to do to be interpreted as the checkpoint 'definition'?

**Comment by Wouter1** [comment]
@wbaccinelli  It seems calling a checkpoint just does not work, I tried to define a checkpoint in my stories file but the story just seems to END at the moment the checkpoint is reached, instead of calling it

**Comment by Wouter1** [comment]
Moved all code to 1 section, no checkpoint.
Now it runs straight over the form action, instead of waiting for the user response.
I re-added the "active-loop" code but rasa seems to exit my story.

Probably due to this
```
rasa_bot-rasa_actions-1  | 2022-10-03 13:01:18 ERROR    rasa_sdk.endpoint  - No registered action found for name 'validate_activity_usefulness_form'.
rasa_bot-rasa_server-1   | 2022-10-03 13:01:18 ERROR    rasa.core.processor  - Encountered an exception while running action 'validate_activity_usefulness_form'.Bot will continue, but the actions events are lost. Please check the logs of your action server for more information.
rasa_bot-rasa_server-1   | Traceback (most recent call last):
rasa_bot-rasa_server-1   |   File "/usr/local/lib/python3.8/site-packages/rasa/core/actions/action.py", line 741, in run
rasa_bot-rasa_server-1   |     response: Any = await self.action_endpoint.request(
rasa_bot-rasa_server-1   |   File "/usr/local/lib/python3.8/site-packages/rasa/utils/endpoints.py", line 173, in request
rasa_bot-rasa_server-1   |     raise ClientResponseError(
rasa_bot-rasa_server-1   | rasa.utils.endpoints.ClientResponseError: 404, Not Found, body='b'{"error":"No registered action found for name \'validate_activity_usefulness_form\'.","action_name":"validate_activity_usefulness_form"}''
rasa_bot-rasa_server-1   | 
rasa_bot-rasa_server-1   | The above exception was the direct cause of the following exception:
rasa_bot-rasa_server-1   | 
rasa_bot-rasa_server-1   | Traceback (most recent call last):
rasa_bot-rasa_server-1   |   File "/usr/local/lib/python3.8/site-packages/rasa/core/processor.py", line 874, in _run_action
rasa_bot-rasa_server-1   |     events = await action.run(
rasa_bot-rasa_server-1   |   File "/usr/local/lib/python3.8/site-packages/rasa/core/actions/action.py", line 765, in run
rasa_bot-rasa_server-1   |     raise RasaException(
rasa_bot-rasa_server-1   | rasa.shared.exceptions.RasaException: Failed to execute custom action 'validate_activity_usefulness_form'

```

**Comment by Wouter1** [comment]
I did not declare the form in the forms section of the domain definition.
Strange, I think other code also uses forms and I did not see a forms section there. 
Edit: in domain_future_self_dialog there IS a forms section.

**Comment by Wouter1** [comment]
Maybe the action is undefined because there is a problem with the python file? 
```
>>> import actions.actions_general_activity
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/wouter/git/virtual-coach-rasa/Rasa_Bot/actions/actions_general_activity.py", line 5, in <module>
    class ValidateActivityUsefullnessForm(FormValidationAction):
  File "/home/wouter/git/virtual-coach-rasa/Rasa_Bot/actions/actions_general_activity.py", line 6, in ValidateActivityUsefullnessForm
    def name(self) -> Text:
NameError: name 'Text' is not defined
```

**Comment by Wouter1** [comment]
enabled  mypy to see these kind of errors in the editor already.

**Comment by Wouter1** [comment]
Fixed the import errors, I can now import the file properly and mypy does not give any errors/warnings on it. But the error persists.

**Comment by Wouter1** [comment]
The error already happens on the ```    - action: validate_activity_usefulness_form``` so this has nothing to do with 'active_loop'.

**Comment by Wouter1** [comment]
The stories_future_self defines  pick_smoker_words_form
However there is no action with this name? I thought that all action calls must be defined in an actions file

**Comment by Wouter1** [comment]
Changed name of the form in domain_general_activity and using that new iname in the stories and removed the actions section . Now it seems fine, but it does not await for a user's response, it just falls through

What is also a big problem is that rasa does not seem to check anything. It just builds (which takes also long) and then fails or misbehave when you run it. This is terrible for finding the issue. I'm basically staring the whole day at the rasa building stuff and stopping/restarting docker

**Comment by Wouter1** [comment]
The validate_activity_usefulness_form is in the actions list and looks like standard action. But it seems never called. 

Similar with validate_see_myself_as_mover_form, and validate_confirm_words_form:
defined as action and the action exists; but it's never called.

**Comment by Wouter1** [comment]
Also it looks like they recently changed the forms and stories formats. I see lots of examples that look like a big syntax error

**Comment by Wouter1** [comment]
Strange problem in eclipse/git. If I type just 1 line of text as commit message, it refuses to commit with "you must enter a commit message". If I add multiple additional lines with nonsense it works fine

**Comment by Wouter1** [comment]
I had forgotten the obligatory response if I want to use the form. Not sure but I never got an error message on that. But added.

```
utter_ask_activity_usefulness_form_activity_useful_rating:
  - text: "Hoe nuttig vond je deze activiteit? Dit kun je aangeven op een schaal van 0-10. 0 is 'helamaal niet nuttig'. 10 is 'heel erg nuttig'. Typ nu een cijfer van 0-10."
```

It seems the story now does not even start the form as I do not get the question to enter the number either. It seems the story is cancelled but not geting any error

**Comment by Wouter1** [comment]
Checked https://www.youtube.com/watch?v=U3an88fjFRc

Still confused.
They give this example
```
-rule: Activate form
 steps:
 - intent: request_names
 - action: name_form
 - active_loop: name_form

- rule: submit form
  condition:
  - active_loop: name_form
  steps:
  - action: name_form
  - active_loop: null
  - slot_was_set: 
    - request_slot: null
  - action: utter_submit
  - action: utter_slots_values
```
The video explains that the first steps in the submit form are CONDITIONS. 
I can not follow, why are actions conditions ? because slot_was_set is a condition? And why are the ones after that actions?

**Comment by Wouter1** [comment]
Discussing  with Walter.
We removed the 'entity' from the stories_
We fixed thedocker-compose file (use "Dockerfile" instead of *.dockerfile). We found this after Walter asked to start docker with an extra --build option, so ```docker compose up --build```, which gave errors

We removed actions_general_activity.py becausse now it does not compile anymore

Now Walter is debugging using ```rasa run --debug```. gives extra info  and checking the logs
The logs show something with "validating activity_useful_rating" even if you request a totally different intent eg from future-self-dialog

**Comment by Wouter1** [comment]
@wbaccinelli  the system seems to get back working as soon as I re-add the entities that we removed. So change the beginning of  domain_general_activity to
```
entities:
- activity_useful_rating


slots:
  # The user's rating of previous activity
  activity_useful_rating:
    type: float
    min_value: 0
    max_value: 10
    initial_value: 0
    mappings:
    - type: from_entity
      entity: activity_useful_rating
```    
It's still not behaving as expected but at the review_last_activity starts

**Comment by Wouter1** [comment]
Seems that if you do ```docker compose up --build``` you get better error messages if python code can not be imported

**Comment by Wouter1** [comment]
Done several fixes so that the dialog at least works again.
The thanks now also mentiones the "received" rating. 
However the system still does not wait for the user to enter the rating

**Comment by wbaccinelli** [comment]
After playing around for some time I am in the same situation. Looking at the debug lines of Rasa, it seems that the `activity_useful_rating` get filled with the `review_last_activity` intent value. I am still trying to figure out why.

**Comment by Wouter1** [comment]
Tried again with the 'entity' based approach, 
```
entities:
- activity_useful_rating


slots:
  # The user's rating of previous activity
  activity_useful_rating:
    type: float
    min_value: 0
    max_value: 10
    initial_value: 0
    mappings:
    - type: from_entity
      action: number
```
But rasa does not even respond to messages with this setting.

**Comment by Wouter1** [comment]
@wbaccinelli  did you find the source of the problems that we found with my code?

**Comment by wbaccinelli** [comment]
@Wouter1 not yet, I am still figuring this out

**Comment by Wouter1** [comment]
@wbaccinelli  can you report the version number of this problematic code? We may have to refer to it for later discussion/bug report but I forgot to note it and I did several commits afterwards.

**Comment by Wouter1** [comment]
We discussed on the sprint meeting.
Bouke has same problems.

But Bouke got something working. It's on the relapse-dialog branch.

Main difference seems that he uses categories instead of text/float
Walter asks if I can also revert to categories for my case

Also I need to not set initial value, and that I need to check that the slot was set after starting the acttive_loop

Also I can use logging.info("fdsfds") instead of dispatcher.utter_message(response="blabla") to log messages from the code.

**Comment by Wouter1** [comment]
But how can you check that a category was set to anything, rather than a specific number?

**Comment by Wouter1** [comment]
@wbaccinelli  @bscheltinga 

I can not use categories because I need to proceed the dialog if the slot is filled, with *anything*.

In that case, I am supposed to use the type "text", not "categories"
see also

https://forum.rasa.com/t/categorical-slots-with-slot-was-set-stories/38116

**Comment by Wouter1** [comment]
Changed the type of activity_useful_rating to text and  mapping of the activity_useful_rating slot to 'from_text'.

**Comment by Wouter1** [comment]
That does not fix the problem. It seems  the form is not activating.

Added "    influence_conversation: true" to slot. 
Does not change anything. I think this is also the default?
Yes it is default true: https://rasa.com/docs/rasa/domain/

**Comment by Wouter1** [comment]
changed mappings from
```
    mappings:
    - type: from_text
      action: activity_usefulness_form

```

 to 
```
    mappings:
    - type: from_text
      conditions:
      - active_loop: activity_usefulness_form
```

**Comment by Wouter1** [comment]
Does not change anything?

**Comment by Wouter1** [comment]
Checked and compared domain actions and stories with Bouke's code.
Seems all similar though I need to use text.
But the validate action is still not triggering. I dont get errors either.

**Comment by bscheltinga** [comment]
I just wanted to say, make sure the validate action is running, I found that it's crucial for the form to work. Maybe you should check the name of the validation actions

**Comment by Wouter1** [comment]
Mailed Bouke if he can help me

**Comment by Wouter1** [comment]
@bscheltinga The action is not running but I dont see why. All names seem ok

**Comment by Wouter1** [comment]
@bscheltinga I now see that you return the name of an *action* in your ValidateSmokeOrPaForm. I thought the name of the form must be there.

**Comment by Wouter1** [comment]
@bscheltinga  @wbaccinelli  yes finally the form is triggering.
I had the name of the validate action and the name of the form identical. I thought that was how it was supposed to be. Now I changed the name of the validate action to validate_.... the form validation is finally called.

**Comment by Wouter1** [comment]
The validate action runs one time. Then it seems I drop out of the form and out of the story, back to the main level:
>me: Vorige Activiteit
>bot: Hoe nuttig vond je deze activiteit? Dit kun je aangeven op een schaal van 0-10. 0 is 'helamaal niet nuttig'. 10 is 'heel erg nuttig'. Typ nu een cijfer van 0-10.
>me: Vorige Activiteit
>bot: Hoe nuttig vond je deze activiteit? Dit kun je aangeven op een schaal van 0-10. 0 is 'helamaal niet nuttig'. 10 is 'heel erg nuttig'. Typ nu een cijfer van 0-10.
>me: 1
>bot:Sorry, ik begrijp (nog) niet wat je bedoelt,...

in rasa logs I see
```
......
rasa_bot-rasa_server-1   | (Background on this error at: https://sqlalche.me/e/14/e3q8).
rasa_bot-rasa_actions-1  | 2022-10-05 11:18:57 INFO     root  - validating activity rating
rasa_bot-rasa_server-1   | 2022-10-05 11:18:58 ERROR    rasa.core.tracker_store  - Error happened when trying to save conversation tracker to 'SQLTrackerStore'. Falling back to use the 'InMemoryTrackerStore'. Please investigate the following error: (sqlite3.OperationalError) unable to open database file
rasa_bot-rasa_server-1   | (Background on this error at: https://sqlalche.me/e/14/e3q8).
rasa_bot-rasa_server-1   | 2022-10-05 11:18:59 ERROR    rasa.core.tracker_store  - Error happened when trying to save conversation tracker to 'SQLTrackerStore'. Falling back to use the 'InMemoryTrackerStore'. Please investigate the following error: (sqlite3.OperationalError) unable to open database file
rasa_bot-rasa_server-1   | (Background on this error at: https://sqlalche.me/e/14/e3q8).
rasa_bot-rasa_server-1   | 2022-10-05 11:18:59 ERROR    rasa.core.tracker_store  - Error happened when trying to save conversation tracker to 'SQLTrackerStore'. Falling back to use the 'InMemoryTrackerStore'. Please investigate the following error: (sqlite3.OperationalError) unable to open database file
rasa_bot-rasa_server-1   | (Background on this error at: https://sqlalche.me/e/14/e3q8).
rasa_bot-rasa_actions-1  | 2022-10-05 11:18:59 INFO     root  - validating activity rating
rasa_bot-rasa_server-1   | 2022-10-05 11:18:59 ERROR    rasa.core.tracker_store  - Error happened when trying to save conversation tracker to 'SQLTrackerStore'. Falling back to use the 'InMemoryTrackerStore'. Please investigate the following error: (sqlite3.OperationalError) unable to open database file
rasa_bot-rasa_server-1   | (Background on this error at: https://sqlalche.me/e/14/e3q8).



```

**Comment by Wouter1** [comment]
Trying this
```
  - slot_was_set:
    - activity_useful_rating

```

but does not work

**Comment by Wouter1** [comment]
@bscheltinga The validation runs, but then I seem to drop out of the story. 

This is the story
```
- story: introduction PF
  steps:
  - intent: review_last_activity
  - action: utter_checking_activities
  - action: activity_usefulness_form
  - active_loop: activity_usefulness_form
  - slot_was_set:
    - activity_useful_rating: ja
  - action: utter_ok
```
and the story seems dropped around active_loop, maybe in slot_was_set

**Comment by Wouter1** [comment]
Trying to run rasa in debug, with this as last line in Dockerfile
```CMD ["rasa", "run", "--enable-api", "--debug"]```

**Comment by Wouter1** [comment]
@wbaccinelli I seem to drop out of the story after the call to the validate action. I dont see unusual errors, I see this in rasa if I turn on debug mode
```
rasa_bot-rasa_server-1   | [state 3] user intent: review_last_activity | previous action name: activity_usefulness_form | active loop: {'name': 'activity_usefulness_form'} | slots: {'current_intervention_component': (1.0,)}
rasa_bot-rasa_server-1   | 2022-10-05 11:50:56 DEBUG    rasa.core.policies.memoization  - There is no memorised next action
rasa_bot-rasa_server-1   | 2022-10-05 11:50:56 DEBUG    rasa.engine.graph  - Node 'run_RulePolicy1' running 'RulePolicy.predict_action_probabilities'.
rasa_bot-rasa_server-1   | 2022-10-05 11:50:56 DEBUG    rasa.core.policies.rule_policy  - Predicted 'action_listen' after loop 'activity_usefulness_form'.
rasa_bot-rasa_server-1   | 2022-10-05 11:50:56 DEBUG    rasa.engine.graph  - Node 'select_prediction' running 'DefaultPolicyPredictionEnsemble.combine_predictions_from_kwargs'.
rasa_bot-rasa_server-1   | 2022-10-05 11:50:56 DEBUG    rasa.core.policies.ensemble  - Predicted next action using RulePolicy.
rasa_bot-rasa_server-1   | 2022-10-05 11:50:56 DEBUG    rasa.core.processor  - Predicted next action 'action_listen' with confidence 1.00.
rasa_bot-rasa_server-1   | 2022-10-05 11:50:56 DEBUG    rasa.core.processor  - Policy prediction ended with events '[]'.
rasa_bot-rasa_server-1   | 2022-10-05 11:50:56 DEBUG    rasa.core.processor  - Action 'action_listen' ended with events '[]'.
```


And, I see this "  - EXTERNAL_trigger_relapse_phase" intent in the domain_relapse_dialog. Is this needed for something?

**Comment by Wouter1** [comment]
@wbaccinelli is this correct behaviour ?
>ik:EHBO doos
>bot: Sorry, ik begrijp (nog) niet wat je bedoelt, kun je me het op een andere manier duidelijk maken?
>ik:Get first aid kit
>bot: Sorry, ik begrijp (nog) niet wat je bedoelt, kun je me het op een andere manier duidelijk maken?

**Comment by Wouter1** [comment]
Walter replied

>Hi Bouke and Wouter,

 

>I wrote this same thing on the Teams channel, but I realized that maybe an email would have been more convenient.

 

>I found a solution to prevent the form to fill the slot with the previous intent content. In the form declaration in the domain file you have to add the ignored_intents field and list the intents you want to be ignored.

 

>Here attached, the domain and story file that are working for me.

domain:
 ```
intents:
- review_last_activity

slots:
  activity_useful_rating:
    type: float
    influence_conversation: false
    mappings:
      - type: from_text
        conditions:
        - active_loop: activity_usefulness_form

responses:
  utter_checking_activities:
  - text: "Laten we eerst terug kijken op je vorige activiteit."
  utter_ok:
  - text: "ok thanks for your rating"
  utter_bye:
  - text: "bye!"

  # form slot responses obligatory see https://rasa.com/docs/rasa/forms/
  utter_ask_activity_usefulness_form_activity_useful_rating:
  - text: "Hoe nuttig vond je deze activiteit? Dit kun je aangeven op een schaal van 0-10. 0 is 'helamaal niet nuttig'. 10 is 'heel erg nuttig'. Typ nu een cijfer van 0-10."

actions:
  - action_reset_activity_useful_rating

forms:
  activity_usefulness_form:
    ignored_intents:
      - review_last_activity
    required_slots:
      - activity_useful_rating
```

and stories:
```
stories:
- story: introduction PF
  steps:
    - intent: review_last_activity
    - action: utter_checking_activities
    - action: activity_usefulness_form
    - active_loop: activity_usefulness_form
    - active_loop: null
    - action: utter_ok
    - ```

**Comment by Wouter1** [comment]
@wbaccinelli  @bscheltinga 
I'm usually not on teams.

**Comment by Wouter1** [comment]
@wbaccinelli  I added ignored_intents to the activity_usefulness_form
Notice that the domain file was already changed a lot but I added it anyways
But it does not seem to help, the story seems to end after or somewhere during the form, I do not get out of the form and the slot remains containing  'None'.

**Comment by wbaccinelli** [comment]
@Wouter1 can you go back to the working domain and story structure I sent? The implementation of the dialog is still at that stage, from what I see, so I think that spending time on fixing a different structure is pointless.

**Comment by Wouter1** [comment]
@bscheltinga  @wbaccinelli  which working domain are you exactly referring to (which version nr)
I did the changes to make the code match Bouke's code, and later I found another issue (name returned by the form had to be different).

**Comment by Wouter1** [comment]
@wbaccinelli Ah now I see, you want me to just replace the domain and stories files with your text. I'll do and report back.

**Comment by Wouter1** [comment]
@wbaccinelli 

Ok this is the bot behaviour after replacing domain and stories with your texts. Thisd is on top of Commit 97ceca7b382fdc77107710c814be605a69a08b86
>me: Vorige Activiteit
>bot:[{"recipient_id":"41538","text":"Laten we eerst terug kijken op je vorige activiteit."},{"recipient_id":"41538","text":"Hoe nuttig vond je deze activiteit? Dit kun je aangeven op een schaal van 0-10. 0 is 'helamaal niet nuttig'. 10 is 'heel erg nuttig'. Typ nu een cijfer van 0-10."}]
>me: 2'
>bot:[{"recipient_id":"41538","text":"Sorry, ik begrijp (nog) niet wat je bedoelt, kun je me het op een andere manier duidelijk maken?"},{"recipient_id":"41538","text":"Je kunt deze woorden gebruiken om met mij te praten: 1) EHBO doos, 2) ..., ... ."}]

**Comment by Wouter1** [comment]
@wbaccinelli @bscheltinga 
I'm trying to figure out what goes wrong in the form 

When I send a message to Rasa (which runs in --debug) I always see this first in the log
```
rasa_bot-rasa_server-1   | 2022-10-05 13:46:25 DEBUG    rasa.core.lock_store  - Issuing ticket for conversation '41538'.
rasa_bot-rasa_server-1   | 2022-10-05 13:46:25 DEBUG    rasa.core.lock_store  - Acquiring lock for conversation '41538'.
rasa_bot-rasa_server-1   | 2022-10-05 13:46:25 DEBUG    rasa.core.lock_store  - Acquired lock for conversation '41538'.
rasa_bot-rasa_server-1   | 2022-10-05 13:46:25 ERROR    rasa.core.tracker_store  - Error happened when trying to save conversation tracker to 'SQLTrackerStore'. Falling back to use the 'InMemoryTrackerStore'. Please investigate the following error: (sqlite3.OperationalError) unable to open database file
rasa_bot-rasa_server-1   | (Background on this error at: https://sqlalche.me/e/14/e3q8).
rasa_bot-rasa_server-1   | 2022-10-05 13:46:25 ERROR    rasa.core.tracker_store  - Error happened when trying to save conversation tracker to 'SQLTrackerStore'. Falling back to use the 'InMemoryTrackerStore'. Please investigate the following error: (sqlite3.OperationalError) unable to open database file
rasa_bot-rasa_server-1   | (Background on this error at: https://sqlalche.me/e/14/e3q8).
rasa_bot-rasa_server-1   | 2022-10-05 13:46:25 DEBUG    rasa.core.processor  - Starting a new session for conversation ID '41538'.
rasa_bot-rasa_server-1   | 2022-10-05 13:46:25 DEBUG    rasa.core.processor  - Policy prediction ended with events '[]'.
rasa_bot-rasa_server-1   | 2022-10-05 13:46:25 DEBUG    rasa.core.processor  - Action 'action_session_start' ended with events '[<rasa.shared.core.events.SessionStarted object at 0x7fd907f23400>, ActionExecuted(action: action_listen, policy: None, confidence: None)]'.
```

The message "Starting a new session" worries me. This session needs to be reconnected, not started anew? Is the session maybe restarted every time I send a new message? That would explain why the form is not working and I get back to the initial state?

**Comment by Wouter1** [comment]
Checked the Rasa source code. 
The message ```Starting a new session for conversation ID``` is indeed logged whenever a session expired, or when when applied_events() returns  None/empty list. Not sure what the last condition exactly means (maybe it means, no actions have been done) but I assume it's similar to deciding that session expired

**Comment by Wouter1** [comment]
mailed walter
Sorry, I'm out of ideas and I dont significant progress. I have (re)read the manuals a few times and seen instruction videos.

Is there any form in the perfectfit system that actually working? How do I get at it?

Because even if I just ask for "ehbo doos" I get "Sorry, ik begrijp  niet ....."

**Comment by Wouter1** [comment]
brief meeting with Bouke
He says that message "future self dialog" should work and get you into a form. If you then enter 'later' then you should get a follow up question on when you want

**Comment by Wouter1** [comment]
Not working with my current version of the 290-general-activity branch. I will test also on main branch

**Comment by Wouter1** [comment]
Removed all changes, switching to main branch for testing there.
This also removes the --debug option in Dockerfile etc

**Comment by Wouter1** [comment]
Again changing docker-compose.yml: remove .rasa_server and .rassa_actions

**Comment by Wouter1** [comment]
On main branch. rasa built anew. docker started anew, debug mode.
This is how "future self dialog" runs here

>me: curl --location --request POST 'http://localhost:5005/webhooks/rest/webhook' --header 'Content-Type text/plain' --data-raw '{ "sender":"41538", "message":"future self dialog" } '


>bot: [{"recipient_id":"41538","text":"Hey NAMENOTSET!"},{"recipient_id":"41538","text":"We gaan nu aan de slag met een oefening om jou te helpen met stoppen met roken en genoeg laten bewegen. ..."},{"recipient_id":"41538","text":"Je gaat ongeveer 15 minuten bezig zijn. ..."},{"recipient_id":"41538","text":"Het is belangrijk dat je de oefening in een keer afmaakt. ..."},{"recipient_id":"41538","text":"Geef alsjeblieft antwoord met 'nu' of 'later.'"},{"recipient_id":"41538","text":"Wil jij deze oefening nu doen of later?"}]

>me: curl --location --request POST 'http://localhost:5005/webhooks/rest/webhook' --header 'Content-Type text/plain' --data-raw '{ "sender":"41538", "message":"later" } '

>bot: [{"recipient_id":"41538","text":"Sorry, ik begrijp (nog) niet wat je bedoelt, kun je me het op een andere manier duidelijk maken?"},{"recipient_id":"41538","text":"Je kunt deze woorden gebruiken om met mij te praten: 1) EHBO doos, 2) ..., ... ."}]wouter@nelesubuntu

**Comment by Wouter1** [comment]
There are changes on the main branch.
Let's try the latest

**Comment by Wouter1** [comment]
Latest behaves the same

**Comment by Wouter1** [comment]
mailed Bouke, Walter
Hee Bouke,Walter

I did a full revert to the main branch. 

Forms seem not to work properly there as well. For example:

    me: curl --location --request POST 'http://localhost:5005/webhooks/rest/webhook' --header 'Content-Type text/plain' --data-raw '{ "sender":"41538", "message":"future self dialog" } '

    bot: [{"recipient_id":"41538","text":"Hey NAMENOTSET!"},{"recipient_id":"41538","text":"We gaan nu aan de slag met een oefening om jou te helpen met stoppen met roken en genoeg laten bewegen. ..."},{"recipient_id":"41538","text":"Je gaat ongeveer 15 minuten bezig zijn. ..."},{"recipient_id":"41538","text":"Het is belangrijk dat je de oefening in een keer afmaakt. ..."},{"recipient_id":"41538","text":"Geef alsjeblieft antwoord met 'nu' of 'later.'"},{"recipient_id":"41538","text":"Wil jij deze oefening nu doen of later?"}]

    me: curl --location --request POST 'http://localhost:5005/webhooks/rest/webhook' --header 'Content-Type text/plain' --data-raw '{ "sender":"41538", "message":"later" } '

    bot: [{"recipient_id":"41538","text":"Sorry, ik begrijp (nog) niet wat je bedoelt, kun je me het op een andere manier duidelijk maken?"},{"recipient_id":"41538","text":"Je kunt deze woorden gebruiken om met mij te praten: 1) EHBO doos, 2) ..., ... ."}]wouter@nelesubuntu


    It looks like rasa completely resets the dialog, even though I am using the same sender ID all the time..

**Comment by Wouter1** [comment]
Checking the nasty details of rasa's REST api .... https://rasa.com/docs/rasa/connectors/your-own-website/#rest-channels

**Comment by Wouter1** [comment]
Informed nele that we never got the rasa to completely work and that therefore I could not do anything

**Comment by Wouter1** [comment]
[260922 C1.16 General Activity Dialogue.xlsx](https://github.com/PerfectFit-project/virtual-coach-issues/files/9724117/260922.C1.16.General.Activity.Dialogue.xlsx)

Here is the dialog to be implemented

**Comment by Wouter1** [comment]
The if-then-else-if-then construct in the table is very confusing. 
I need the graph that I supposedly was sent

**Comment by Wouter1** [comment]
![ga_dialog](https://user-images.githubusercontent.com/7699948/194315578-0224987e-abd4-4456-8b10-cd6dda418be9.png)

**Comment by Wouter1** [comment]
Auh?? That does not match at all with the excel sheet.

**Comment by Wouter1** [comment]
added a graphml object with the correct(I think) flow graph to the git repo (I can not add it here, format not supported)

The grapml file can be opened with yed (free software).

**Comment by Wouter1** [comment]
@wbaccinelli @bscheltinga  
The story flow from the excel sheet is confusing.
And the flow diagram that came with it is a total mismatch
Instead check the general_activity_dialog.graphml file that I put in the data folder (in the 290 branch of course)

**Comment by Wouter1** [comment]
@wbaccinelli  @bscheltinga 
I think I completed the general_activity dialog.
I did not write the actions (no time left) but I did write the intended behaviour of the action in the domain description, as the comment after the action 
I just wrote what I think should work, nothing is really run in Rasa since we never got that working.

Maybe I have a little time next week to attend the meeting but I'm on another project then

**Comment by Wouter1** [comment]
@wbaccinelli just as a note, 
Because we have a separate git for the issues, github can not properly insert the commit messages into the ticket. Normally this ticket would have contained messages at all the moments I committed something automatically , because I always add the ticket number to the commit message and git recognises that. Just a thing to consider if you decide on separating the tickets from the actual project

**Comment by wbaccinelli** [comment]
The issue cannot be closed until the dialog has been fully implemented and the code has been reviewed and merged into the main branch.

**Comment by wbaccinelli** [comment]
Blocked by: #289
